### ISSUE TITLE
Keycloaks keep restarting / never come up properly due to recent  Add startupProbe (#408) 0dca237

### ISSUE THREAD

**Issue by pmorch** [codecentric/helm-charts#420]
After the introduction of 0dca237 to fix #408, we're seeing that after a fresh installation, our keycloaks keep restarting. Here are the events from `kubectl -n keycloak describe pod keycloak-0`:

```
Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  4m15s                  default-scheduler  Successfully assigned keycloak/keycloak-0 to karbon-dev-8804a0-k8s-worker-4
  Normal   Pulled     4m15s                  kubelet            Container image "docker.io/busybox:1.32" already present on machine
  Normal   Created    4m15s                  kubelet            Created container pgchecker
  Normal   Started    4m13s                  kubelet            Started container pgchecker
  Normal   Pulling    2m24s (x2 over 3m52s)  kubelet            Pulling image "harbor.example.com/dev/our-keycloak:develop"
  Normal   Pulled     2m24s (x2 over 3m52s)  kubelet            Successfully pulled image "harbor.example.com/dev/our-keycloak:develop"
  Normal   Created    2m24s (x2 over 3m52s)  kubelet            Created container keycloak
  Normal   Started    2m24s (x2 over 3m51s)  kubelet            Started container keycloak
  Warning  Unhealthy  115s (x6 over 3m45s)   kubelet            Liveness probe failed: Get http://172.22.147.139:8080/auth/: dial tcp 172.22.147.139:8080: connect: connection refused
  Normal   Killing    115s (x2 over 3m25s)   kubelet            Container keycloak failed liveness probe, will be restarted
  Warning  Unhealthy  108s (x3 over 3m18s)   kubelet            Readiness probe failed: Get http://172.22.147.139:8080/auth/realms/master: dial tcp 172.22.147.139:8080: connect: connection refused
  Warning  Unhealthy  97s (x3 over 2m57s)    kubelet            Readiness probe failed: Get http://172.22.147.139:8080/auth/realms/master: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

Here is our `custom-values.yaml`:

```yaml
replicas: 3

podLabels: {monitoring: apps}

image:
  repository: harbor.example.com/dev/our-keycloak
  tag: "develop"
  pullPolicy: Always

extraEnv: |
  - name: KEYCLOAK_USER
    value: admin
  - name: KEYCLOAK_PASSWORD
    value: admin
  - name: PROXY_ADDRESS_FORWARDING
    value: "true"

  - name: JGROUPS_DISCOVERY_PROTOCOL
    value: dns.DNS_PING
  - name: JGROUPS_DISCOVERY_PROPERTIES
    value: 'dns_query={{ include "keycloak.serviceDnsName" . }}'
  - name: CACHE_OWNERS_COUNT
    value: "2"
  - name: CACHE_OWNERS_AUTH_SESSIONS_COUNT
    value: "2"

resources:
  requests:
    cpu: 100m
    memory: 600Mi
  limits:
    cpu: 1000m
    memory: 1500Mi

service:
  labels: {monitoring: apps}

ingress:
  enabled: true
  annotations:
    ingress.kubernetes.io/affinity: "cookie"
    ingress.kubernetes.io/session-cookie-name: "INGRESS_SESSION_ID"
  servicePort: http
  rules:
    - host: 'keycloak.example.com'
      paths:
        - /
  tls:
    - hosts:
        - keycloak.example.com
```

Appending this to  `custom-values.yaml` makes it stop. It was done by a colleague, and I must confess I'm not sure what it does and how that relates to 0dca237 but testing shows that it does the trick:

```yaml
livenessProbe: |
  httpGet:
    path: /auth/
    port: http
  initialDelaySeconds: 300
  timeoutSeconds: 5
```

Needless to say, I expect that using the default values doesn't result in a restart loop.

**Comment by mircohaug** [comment]
Hi @pmorch, you are correct this should not happen. Which Kubernetes version do you use? Startup Probes where first introduced in Kubernetes 1.16 in Sept 2019.

**Comment by SeWieland** [comment]
I'm trying to upgrade from chart ~9 to ~10 currently and am seeing the exact same problem.
However, I think it is not the health probes. Check the logs on the container. 

```keycloak 11:32:17,082 ERROR [org.jboss.as.controller.management-operation[] (Controller Boot Thread) WFLYCTL0013: Operation ("add") failed - address: ([
keycloak     ("subsystem" => "infinispan"),
keycloak     ("cache-container" => "keycloak"),
keycloak     ("thread-pool" => "state-transfer")
keycloak ]) - failure description: {"WFLYCTL0080: Failed services" => {"org.wildfly.clustering.infinispan.cache-container.keycloak" => "org.infinispan.manager.EmbeddedCacheManagerStartupException: org.infinispan.commons.CacheConfigurationException: Error starting component org.infinispan.statetransfer.StateTransferManager
keycloak     Caused by: org.infinispan.manager.EmbeddedCacheManagerStartupException: org.infinispan.commons.CacheConfigurationException: Error starting component org.infinispan.statetransfer.StateTransferManager
keycloak     Caused by: org.infinispan.commons.CacheConfigurationException: Error starting component org.infinispan.statetransfer.StateTransferManager
keycloak     Caused by: java.util.concurrent.CompletionException: org.infinispan.remoting.RemoteException: ISPN000217: Received exception from keycloak-0, see cause for remote stack trace
keycloak     Caused by: org.infinispan.remoting.RemoteException: ISPN000217: Received exception from keycloak-0, see cause for remote stack trace
keycloak     Caused by: org.infinispan.commons.CacheException: Unknown command id 85!"}}
``` 

On the other container in my StatefulSet I get:

```
keycloak 11:36:44,387 ERROR [org.infinispan.CLUSTER[] (thread-39,ejb,keycloak-0) ISPN000474: Error processing request 1259@keycloak-1: org.infinispan.commons.CacheException: Unknown command id 85!
```

Guess we have to "scale down to one" first?

**Comment by pmorch** [comment]
@mircohaug here are the versions:


```
$ kubectl --context ourcluster version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.13", GitCommit:"30d651da517185653e34e7ab99a792be6a3d9495", GitTreeState:"clean", BuildDate:"2020-10-15T00:59:17Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by SeWieland** [comment]
Seems to be related. https://github.com/codecentric/helm-charts/issues/50
Rolling updates seem not to be supported. A note in the docs about that in the docs would be helpful I guess :)

So sclaing to 0 and up again resolved all issues on my side. 👍

**Comment by mircohaug** [comment]
@SeWieland where would you expect this notice in the docs?

@pmorch this seems as it should work... i will try to reproduce with the versions you provided and come back to you.

**Comment by mircohaug** [comment]
hey @pmorch I looked further into this and I could not reproduce this at first, but then i noticed your pretty low requested CPU.

I set the limit to the same value and then the pod was not fast enough to startup within the configured timeframe. (~5min) Kubernetes only "guarantees" the requested resources and I suspect, that your pod runs on a cluster where the is currently not enough CPU available to bust beyond the requested 100m. 

To validate this you could watch the container logs before it gets terminated and see if there is a message like `[org.jboss.as] (Controller Boot Thread) WFLYSRV0025: Keycloak 12.0.4 (WildFly Core 13.0.3.Final) started in 64295ms - Started 687 of 973 services (687 services are lazy, passive or on-demand)` before the container is being shut down. If this line is not present the container had not enough CPU available during its startup to startup fast enough.

**Comment by pmorch** [comment]
Hi @mircohaug,

I'm pretty new to Kubernetes. And *especially* the `resources:` seems a black art to me. I still have no idea what to set here.

It just so happens that this was the very first application we published in our rather beefy cluster, so when this occurred, it was the only thing running in the 3-node cluster (apart from the standard k8s stuff, and some prometheus and ELK for logging).

Later, when all was up and running fine, and with > 1000 happy customers, we only see approx. 5% CPU usage in the entire cluster.

I'll talk to our Operations guys about the perhaps sucky `resources:` I've configured.

I will say that it only started happening after the introduction of 0dca237 to fix #408. Worked fine before. But I'll talk to someone with more k8s foo then I :-)

Thanks for the feedback!

Peter

**Comment by pmorch** [comment]
@mircohaug writes:

> I set the limit to the same value and then the pod was not fast enough to startup within the configured timeframe. (~5min)

But, uh... The timeframe is *not* 5min but **30s**. Here the relevant parts from `.spec.containers[0]` from `kubectl get -o yaml pod keycloak-0`:

```
livenessProbe:
  failureThreshold: 3
  httpGet:
    path: /auth/
    port: http
    scheme: HTTP
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
readinessProbe:
  failureThreshold: 3
  httpGet:
    path: /auth/realms/master
    port: http
    scheme: HTTP
  initialDelaySeconds: 30
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 1
```

If I look at the events for keycloak 0:
```
Jun 02 18:37:47 0s          Normal    Started            pod/keycloak-0                    Started container keycloak
Jun 02 18:37:57 0s          Warning   Unhealthy          pod/keycloak-0                    Liveness probe failed: Get http://172.22.147.166:8080/auth/: dial tcp 172.22.147.166:8080: connect: connection refused
Jun 02 18:38:07 0s          Warning   Unhealthy          pod/keycloak-0                    Liveness probe failed: Get http://172.22.147.166:8080/auth/: dial tcp 172.22.147.166:8080: connect: connection refused
Jun 02 18:38:17 0s          Warning   Unhealthy          pod/keycloak-0                    Liveness probe failed: Get http://172.22.147.166:8080/auth/: dial tcp 172.22.147.166:8080: connect: connection refused
Jun 02 18:38:17 0s          Normal    Killing            pod/keycloak-0                    Container keycloak failed liveness probe, will be restarted
```

So 30 seconds after it has been started, Kubernetes decides to restart the container because the liveness probe isn't ready in 30 seconds.

I've also tried beefing up the resources:

```
$ git diff
diff --git a/keycloak-values.yaml b/keycloak-values.yaml
index f4d0a53..7706f43 100644
--- a/keycloak-values.yaml
+++ b/keycloak-values.yaml
@@ -56,10 +56,10 @@ extraEnv: |
 resources:
   requests:
-    cpu: 100m
+    cpu: 1
     memory: 600Mi
   limits:
-    cpu: 1000m
+    cpu: 1.2
     memory: 1500Mi
 
 service:
 ```

So request: 1 CPU and limit 1.2 CPU. And the pods are still rebooting because the pods still aren't ready after 30s.

**Comment by pmorch** [comment]
I haven't been able to find a way for a helm chart to check for whether a feature-gate is enabled, but more than likely your helm foo is superior to mine and perhaps there is a way to test for `startupProbe` support.

So feel free to close this issue, if you deem that because 1.17 is end-of-life anyway, this is a non-issue. But for lost souls like us on 1.17, this is at least why it happens.

**Comment by mircohaug** [comment]
Thanks for coming back with the likely root cause. I don't know of any way to check this either... The startupProbes improves the usability for all supported Kubernetes Versions so maintaining backward compatibility to 1.17 will not be possible. Take care and may you be able to upgrade in the foreseeable future.

**Comment by SeWieland** [comment]
@mircohaug sorry for responding so late.

At least for me, the "root cause" were not the health probes.
As I already mentioned, I think rolling updates are not supported at all (in clusterd mode) due to infinispan caching issues (also mentioned [here](https://www.keycloak.org//docs/latest/upgrading/index.html#running-the-domain-clustered-mode-upgrade-script))

So for me the only possible way was to scale to 0 (downtime!) and upgrade thereafter.

The [docs](https://github.com/codecentric/helm-charts/tree/master/charts/keycloak#from-chart--1000) currently state:

> The upgrade should be seemless. No special care has to be taken.

This was not the case for me, as I had to deal with (very little) downtime.

I'm not aware of a better solution. Maybe just set cache owners and replicas back to 1 before upgrading could avoid downtime, however I'm not sure about that.

**Comment by andreb89** [comment]
For more information, you should check the Keycloak Jira project. Rolling updates are currently not supported. The only supported option is full stop and start, which will cause a downtime. So you will have to use Recreate rather than RollingUpdate strategy. 

The main tracking Jira issue for the zero-downtime Keycloak upgrade: https://issues.redhat.com/browse/KEYCLOAK-7301

**Comment by pmorch** [comment]
@SeWieland : While I don't doubt you have a problem and the symptoms may be identical, this issue is specifically about a change in behaviour introduced by commit 0dca237 to fix #408. I'm pretty sure that commit only affects health probes. So I suggest that if discussion is to continue about your issue, create a new issue to track the separate and different issue root cause you are experiencing the same symptoms of.
